mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/TaskExecutor.java

```diff
@@ -425,19 +446,41 @@ public CompletableFuture<Ack> submitTask(ExecuteStageRequest request) {
         WrappedExecuteStageRequest wrappedRequest =
             new WrappedExecuteStageRequest(PublishSubject.create(), request);
 
-        Task task = new Task(
-            wrappedRequest,
-            workerConfiguration,
-            masterMonitor,
-            classLoaderHandle,
-            subscriptionStateHandlerFactory,
-            Optional.of(getHostname()),
-            Optional.empty());
+        // do not wait for task processing (e.g. artifact download).
+        getIOExecutor().execute(() -> this.prepareTask(wrappedRequest));
+        return CompletableFuture.completedFuture(Ack.getInstance());
+    }
+
+    private void prepareTask(WrappedExecuteStageRequest wrappedRequest) {
+        try {
+            UserCodeClassLoader userCodeClassLoader = this.taskFactory.getUserCodeClassLoader(
+                wrappedRequest.getRequest(), classLoaderHandle);
+            ClassLoader cl = userCodeClassLoader.asClassLoader();
+            // There should only be 1 task implementation provided by mantis-server-worker.
+            ITask task = this.taskFactory.getITaskInstance(wrappedRequest.getRequest(), cl);
+
+            task.initialize(
+                wrappedRequest,
+                workerConfiguration,
+                masterMonitor,
+                userCodeClassLoader,
+                subscriptionStateHandlerFactory,
+                Optional.of(getHostname())
+            );
+
+            scheduleRunAsync(() -> {
+                setCurrentTask(task);
```

Comment:
setCurrentTask(task) should be run on the main thread right?

Reply:
yeah and that's exactly what "scheduleRunAsync" does (schedule task in main thread).